sycl/test/basic_tests/image/image_read.cpp

```diff
@@ -0,0 +1,251 @@
+  if (resultData.g() != expectedColour.g())
+    correct = false;
+  if (resultData.b() != expectedColour.b())
+    correct = false;
+  if (resultData.a() != expectedColour.a())
+    correct = false;
+  return correct;
+}
+
+template <typename dataT, typename coordT, s::image_channel_type channelType>
+bool test1d(coordT coord, dataT expectedResult) {
+  dataT hostPtr[3];
+  for (int i = 0; i < 3; i++)
+    hostPtr[i] = dataT(0 + i, 20 + i, 40 + i, 60 + i);
+  return test1d_coord<dataT, coordT, channelType>(hostPtr, coord, expectedResult);
+}
+
+template <typename dataT, typename coordT, s::image_channel_type channelType>
+bool test2d(coordT coord, dataT expectedResult) {
+  dataT hostPtr[9];
+  for (int i = 0; i < 9; i++)
+    hostPtr[i] = dataT(0 + i, 20 + i, 40 + i, 60 + i);
+  return test2d_coord<dataT, coordT, channelType>(hostPtr, coord, expectedResult);
+}
+
+template <typename dataT, typename coordT, s::image_channel_type channelType>
+bool test3d(coordT coord, dataT expectedResult) {
+  dataT hostPtr[27];
+  for (int i = 0; i < 27; i++)
+    hostPtr[i] = dataT(0 + i, 20 + i, 40 + i, 60 + i);
+  return test3d_coord<dataT, coordT, channelType>(hostPtr, coord, expectedResult);
+}
+
+template <typename dataT, s::image_channel_type channelType>
+bool test() {
+  bool passed = true;
+
+  // 1d image tests
+  if (!test1d<dataT, int, channelType>(0, dataT(0, 20, 40, 60)))
+    passed = false;
+  if (!test1d<dataT, int, channelType>(1, dataT(1, 21, 41, 61)))
+    passed = false;
+  if (!test1d<dataT, int, channelType>(2, dataT(2, 22, 42, 62)))
+    passed = false;
+
+  // 2d image tests
+  if (!test2d<dataT, s::int2, channelType>(s::int2(0, 0), dataT(0, 20, 40, 60)))
+    passed = false;
+  if (!test2d<dataT, s::int2, channelType>(s::int2(1, 0), dataT(1, 21, 41, 61)))
+    passed = false;
+  if (!test2d<dataT, s::int2, channelType>(s::int2(0, 1), dataT(3, 23, 43, 63)))
+    passed = false;
+  if (!test2d<dataT, s::int2, channelType>(s::int2(1, 1), dataT(4, 24, 44, 64)))
+    passed = false;
+
+  // 3d image tests
+  if (!test3d<dataT, s::int4, channelType>(
+          s::int4(0, 0, 0, 0), dataT(0, 20, 40, 60)))
+    passed = false;
+  if (!test3d<dataT, s::int4, channelType>(
+          s::int4(1, 0, 0, 0), dataT(1, 21, 41, 61)))
+    passed = false;
+  if (!test3d<dataT, s::int4, channelType>(
+          s::int4(0, 1, 0, 0), dataT(3, 23, 43, 63)))
+    passed = false;
+  if (!test3d<dataT, s::int4, channelType>(
+          s::int4(1, 1, 0, 0), dataT(4, 24, 44, 64)))
+    passed = false;
+  if (!test3d<dataT, s::int4, channelType>(
+          s::int4(1, 0, 1, 0), dataT(10, 30, 50, 70)))
+    passed = false;
+  if (!test3d<dataT, s::int4, channelType>(
+          s::int4(0, 1, 1, 0), dataT(12, 32, 52, 72)))
+    passed = false;
+  if (!test3d<dataT, s::int4, channelType>(
+          s::int4(1, 1, 1, 0), dataT(13, 33, 53, 73)))
+    passed = false;
+
+  return passed;
+}
+
+int main() {
+
+  bool passed = true;
+
+  // Half image
```

Comment:
I think we should check that this feature is supported by the device before running the test. Currently it fails on Intel OpenCL CPU and I assume it's because OpenCL CPU doesn't support "half image" feature.

Reply:
Done!